File: app/models/freight.rb
class Freight < ActiveRecord::Base
  belongs_to :car
  belongs_to :client, polymorphic: true

  before_save :freight_belongs_to_client

  def total
    discount_value = (discount || 0) * -1
    [subtotal, ferry, platform, redispatching, platform_origin, platform_destination, discount_value].compact.sum
  end

  private
  def freight_belongs_to_client
    self.client = self.car.client
  end
end

File: app/assets/javascripts/angular/models/freight.js
(function(angular) {
  "use strict";

  angular.module("models.freight", [])

  .factory("Freight", function() {
    function Freight(params) {
      if (!params) params = {};

      this.subtotal = params.subtotal;
      this.ferry = params.ferry;
      this.platform = params.platform;
      this.redispatching = params.redispatching;
      this.platformOrigin = params.platformOrigin;
      this.platformDestination = params.platformDestination;
      this.discount = params.discount;
      this.observation = params.observation;
    }

    Freight.fromJson = function(json) {
      return new Freight({
        subtotal: json.subtotal,
        ferry: json.ferry,
        platform: json.platform,
        redispatching: json.redispatching,
        platformOrigin: json.plaform_origin,
        platformDestination: json.platform_destination,
        discount: json.discount,
        observation: json.observation
      });
    };

    return Freight;
  });
})(window.angular);
